### ISSUE TITLE
how to open this url?

### ISSUE THREAD

**Issue by tubeuchiha** [hierynomus/smbj#483]
how to open like this smb url:    smb://192.168.0.243/nas_nas_nas/nas1/E/369/17869/19257

**Comment by ecki** [comment]
192.. is the host address, nas_nas_nas is the share name and the rest (/nas1/E/../19257) is the path. You need to split the components yourself.

    try (Connection connection = client.connect("192.168.0.243")) {
        AuthenticationContext ac = new AuthenticationContext("USERNAME", "PASSWORD".toCharArray(), "DOMAIN");
        Session session = connection.authenticate(ac);
        try (DiskShare share = (DiskShare) session.connectShare("nas_nas_nas")) {
            Set<SMB2ShareAccess> s = new HashSet<>();
            s.add(SMB2ShareAccess.ALL.iterator().next());
            File f = share.openFile(path, EnumSet.of(AccessMask.GENERIC_READ), null, s, null, null);

**Comment by tubeuchiha** [comment]
> 192.. is the host address, nas_nas_nas is the share name and the rest (/nas1/E/../19257) is the path. You need to split the components yourself.
> 
> ```
> try (Connection connection = client.connect("192.168.0.243")) {
>     AuthenticationContext ac = new AuthenticationContext("USERNAME", "PASSWORD".toCharArray(), "DOMAIN");
>     Session session = connection.authenticate(ac);
>     try (DiskShare share = (DiskShare) session.connectShare("nas_nas_nas")) {
>         Set<SMB2ShareAccess> s = new HashSet<>();
>         s.add(SMB2ShareAccess.ALL.iterator().next());
>         String path = “/nas1/E/369/17869/19257“;
>         File f = share.openFile(path, EnumSet.of(AccessMask.GENERIC_READ), null, s, null, null);
> ```



> 192.. is the host address, nas_nas_nas is the share name and the rest (/nas1/E/../19257) is the path. You need to split the components yourself.
> 
> ```
> try (Connection connection = client.connect("192.168.0.243")) {
>     AuthenticationContext ac = new AuthenticationContext("USERNAME", "PASSWORD".toCharArray(), "DOMAIN");
>     Session session = connection.authenticate(ac);
>     try (DiskShare share = (DiskShare) session.connectShare("nas_nas_nas")) {
>         Set<SMB2ShareAccess> s = new HashSet<>();
>         s.add(SMB2ShareAccess.ALL.iterator().next());
>         String path = “/nas1/E/369/17869/19257“;
>         File f = share.openFile(path, EnumSet.of(AccessMask.GENERIC_READ), null, s, null, null);
> ```

ok, thanks you @ecki 。now i have a  problem : does smbj  support random access to the file? I didn't find relevant API 。if support , the example is ?

**Comment by ecki** [comment]
Yes you can use `f.read(byte[] buffer, long fileOffset, int offset, int length)` with an offset.

**Comment by tubeuchiha** [comment]
> Yes you can use `f.read(byte[] buffer, long fileOffset, int offset, int length)` with an offset.

how to play a video file ?  use " f.read(byte[] buffer, long fileOffset, int offset, int length)" ,but play fail

**Comment by ecki** [comment]
There is too little information and too broad of a question to answer. What fails? What player? Do you download first? Why do you need random access? Need a stream class?

**Comment by tubeuchiha** [comment]
> Yes you can use `f.read(byte[] buffer, long fileOffset, int offset, int length)` with an offset.

can SMBClient use a singleton？a Connection  can have more than one Session? SMBClient, Connection,Session..  What is their relationship？

**Comment by tubeuchiha** [comment]
> There is too little information and too broad of a question to answer. What fails? What player? Do you download first? Why do you need random access? Need a stream class? If you need a `java.io.File` you need to download it into a temporary file first..

My understanding is : one SMBClient, one Connection, more than one Session.but i don't know right or not

**Comment by hierynomus** [comment]
One client, multiple connections per client, multiple sessions per connection, multiple shares per session, multiple files per share...

Up to the File level everything should be thread safe.

**Comment by tubeuchiha** [comment]
> One client, multiple connections per client, multiple sessions per connection, multiple shares per session, multiple files per share...
> 
> Up to the File level everything should be thread safe.

ok, by now  have a error: AndroidRuntime: FATAL EXCEPTION: Thread-42 case by
![image](https://user-images.githubusercontent.com/8343965/60938318-2b1d9100-a306-11e9-8dec-6fcf847fdc4b.png)
I found that in the long run will be the bug

**Comment by ecki** [comment]
This means, in the short run :) that the server did not like the settings. You might find more information in the server log. Did you turn on negotiation? Does the server speak smb2?

**Comment by tubeuchiha** [comment]
> Did you turn on negotiation? Does the server speak smb2?

this is my config :
SmbConfig config = SmbConfig.builder()
                .withTimeout(120, TimeUnit.SECONDS)
                .withSoTimeout(180, TimeUnit.SECONDS)
                .withDfsEnabled(true)
                .withMultiProtocolNegotiate(true)
                .build();
        smbClient=new SMBClient(config);

the smbclient is a singleton. i use it to play a video file from samba server .
